File: examples/go/tax_remittance.go
// Example: Tax Remittance
//
// Remits tax payments to the Kenya Revenue Authority (KRA) via M-Pesa.
// Result arrives via ResultURL.
//
// Requires: MPESA_CONSUMER_KEY, MPESA_CONSUMER_SECRET,
//
//	MPESA_INITIATOR_NAME, MPESA_INITIATOR_PASSWORD
//
// Verified against the sandbox.
package main

import (
	"context"
	"fmt"
	"log"
	"os"

	"github.com/yourdudeken/daraja/sdks/go/client"
	"github.com/yourdudeken/daraja/sdks/go/types"
)

func main() {
	mpesa := client.NewClient(types.MpesaConfig{
		ConsumerKey:       os.Getenv("MPESA_CONSUMER_KEY"),
		ConsumerSecret:    os.Getenv("MPESA_CONSUMER_SECRET"),
		Environment:       types.Sandbox,
		InitiatorName:     os.Getenv("MPESA_INITIATOR_NAME"),
		InitiatorPassword: os.Getenv("MPESA_INITIATOR_PASSWORD"),
	})

	ctx := context.Background()

	resp, err := mpesa.TaxRemittance(ctx, types.TaxRemittanceRequest{
		CommandID:              "PayTaxToKRA",
		SenderIdentifierType:   "4",
		RecieverIdentifierType: "4",
		Amount:                 "100",
		PartyA:                 "174379",
		PartyB:                 "572572", // KRA paybill
		AccountReference:       "TAX-TEST",
		Remarks:                "Test tax remittance",
		QueueTimeOutURL:        "https://example.com/tax/queue",
		ResultURL:              "https://example.com/tax/result",
	})
	if err != nil {
		log.Fatalf("Tax Remittance failed: %v", err)
	}

	fmt.Printf("OriginatorConversationID: %s\n", resp.OriginatorConversationID)
	fmt.Printf("ResponseCode: %s\n", resp.ResponseCode)
}

File: mcp/src/tools/tax-remittance.ts
import type { Mpesa } from "daraja-sdk-ts";
import type { Tool } from "./index.js";

export const taxRemittanceTool: Tool = {
  name: "tax_remittance",
  description: "Remit tax to KRA via M-Pesa. CommandID defaults to PayTaxToKRA. PartyB defaults to 572572 (KRA paybill).",
  required: ["amount", "partyA", "remarks", "resultURL", "queueTimeOutURL"],
  inputSchema: {
    amount: { type: "number", description: "Tax amount to remit" },
    partyA: { type: "string", description: "Payer's shortcode" },
    remarks: { type: "string", description: "Transaction remarks" },
    resultURL: { type: "string", description: "URL for result callback" },
    queueTimeOutURL: { type: "string", description: "URL for timeout callback" },
  },
  handler: async (input, client) => {
    return client.taxRemittance.remit({
      Amount: input.amount as number,
      PartyA: input.partyA as string,
      Remarks: input.remarks as string,
      ResultURL: input.resultURL as string,
      QueueTimeOutURL: input.queueTimeOutURL as string,
      CommandID: "PayTaxToKRA",
      PartyB: "572572",
    } as never);
  },
};
